agent/containermetadata/metadata_manager.go

```diff
@@ -0,0 +1,186 @@
+	// Do nothing if disabled
+	if manager == nil || !manager.enabled {
+		return
+	}
+	manager.containerInstanceARN = containerInstanceARN
+}
+
+// createmetadata creates the metadata file and adds the metadata directory to
+// the container's mounted host volumes
+// Pointer hostConfig is modified directly so there is risk of concurrency errors.
+func (manager *metadataManager) CreateMetadata(hostConfig *docker.HostConfig, task *api.Task, container *api.Container) error {
+	// Do nothing if disabled
+	if manager == nil || !manager.enabled {
+		return nil
+	}
+
+	// Do not create metadata file for internal containers
+	if container.IsInternal {
+		return nil
+	}
+
+	// Create task and container directories if they do not yet exist
+	metadataDirectoryPath, err := getMetadataFilePath(task, container, manager.dataDir)
+	// Stop metadata creation if path is malformed for any reason
+	if err != nil {
+		return fmt.Errorf("container metadata create: %v", err)
+	}
+
+	err = os.MkdirAll(metadataDirectoryPath, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	// Create metadata file
+	err = createMetadataFile(metadataDirectoryPath)
+	if err != nil {
+		return err
+	}
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := manager.parseMetadataAtContainerCreate(task, container)
+	err = metadata.writeToMetadataFile(task, container, manager.dataDir)
+	if err != nil {
+		return err
+	}
+
+	// Add the directory of this container's metadata to the container's mount binds
+	// We do this at the end so that we only mount the directory if there are no errors
+	// This is the only operating system specific point here, so it would be nice if there
+	// were some elegant way to do this for both windows and linux at the same time
+	binds := createBinds(hostConfig.Binds, manager.dataDirOnHost, metadataDirectoryPath, mountPoint, container.Name)
+	hostConfig.Binds = binds
+	return nil
+}
+
+// UpdateMetadata updates the metadata file after container starts and dynamic
+// metadata is available
+func (manager *metadataManager) UpdateMetadata(dockerID string, task *api.Task, container *api.Container) error {
+	// Do nothing if disabled
+	if manager == nil || !manager.enabled {
+		return nil
+	}
+
+	// Do not update (non-existent) metadata file for internal containers
+	if container.IsInternal {
+		return nil
+	}
+
+	// Verify metadata file exists before proceeding
+	if !metadataFileExists(task, container, manager.dataDir) {
+		expectedPath, _ := getMetadataFilePath(task, container, manager.dataDir)
+		return fmt.Errorf("container metadata update: %s does not exist", expectedPath)
+	}
+
+	// Get docker container information through api call
+	dockerContainer, err := manager.client.InspectContainer(dockerID, inspectContainerTimeout)
+	if err != nil {
+		return err
+	}
+
+	// Ensure we do not update a container that is invalid or is not running
+	if dockerContainer == nil || !dockerContainer.State.Running {
```

Comment:
Admittedly, the first condition is added because of paranoia. You are right that if the container is empty we should have gotten an error earlier.